tools/dev-server.js

```diff
@@ -13,34 +13,69 @@ const { Builder } = require("./builder");
 const KARMA_PORT = 9876;
 const SERVE_PORT = 5000;
 
+class KarmaServer {
+  /**
+   * @param {string} configFile
+   * @param {string} [browser]
+   * @param {string} [grep]
+   */
+  constructor(configFile, browser, grep) {
+    this._configFile = configFile;
+    this._browsers = browser ? [browser] : undefined;
+    this._grep = grep;
+    this._isActive = null;
+  }
+
+  start() {
+    this._karmaConfig = karma.config.parseConfig(this._configFile, {
+      browsers: this._browsers,
+      autoWatch: false,
+      port: KARMA_PORT,
+      logLevel: karma.constants.LOG_WARN,
+      client: {
+        args: ["--grep", this._grep || ""],
+      },
+      mochaReporter: { ignoreSkipped: true },
+    });
+    this.karmaServer = new karma.Server(this._karmaConfig);
+    this.karmaServer.start();
+    return new Promise(resolve =>
+      this.karmaServer.once("browsers_ready", resolve)
+    );
+  }
+
+  async stop() {
+    return new Promise(resolve =>
+      karma.stopper.stop(this._karmaConfig, code =>
+        setTimeout(() => resolve(code), 0)
+      )
+    );
+  }
+
+  async run() {
+    if (this._isActive) return;
+    this._isActive = true;
+    karma.runner.run(this._karmaConfig, () => {});
+    await new Promise(res => this.karmaServer.once("run_complete", res));
```

Comment:
Just for consistency...
```suggestion
    await new Promise(resolve => this.karmaServer.once("run_complete", resolve));
```